File: scripts/generate-skills-index.js
#!/usr/bin/env node

/**
 * Generate skills_index.json from all SKILL.md files
 * Extracts YAML frontmatter from skills across all platforms
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

console.log('🔍 Scanning skills...');

const skillsDir = path.join(__dirname, '../.github/skills');
const skills = [];

// Read each skill's SKILL.md
const skillDirs = fs.readdirSync(skillsDir).filter(f => {
  const fullPath = path.join(skillsDir, f);
  return fs.statSync(fullPath).isDirectory();
});

skillDirs.forEach(skillName => {
  const skillPath = path.join(skillsDir, skillName, 'SKILL.md');
  
  if (!fs.existsSync(skillPath)) {
    console.warn(`⚠️  ${skillName}/SKILL.md not found`);
    return;
  }
  
  const content = fs.readFileSync(skillPath, 'utf8');
  
  // Extract YAML frontmatter
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    console.warn(`⚠️  No YAML frontmatter in ${skillName}/SKILL.md`);
    return;
  }
  
  try {
    const metadata = yaml.load(match[1]);
    
    // Validate required fields
    const required = ['name', 'version', 'description', 'category', 'tags', 'risk', 'platforms'];
    const missing = required.filter(field => !metadata[field]);
    
    if (missing.length > 0) {
      console.warn(`⚠️  ${skillName} missing fields: ${missing.join(', ')}`);
    }
    
    skills.push({
      name: metadata.name || skillName,
      version: metadata.version || '0.0.0',
      description: metadata.description || '',
      category: metadata.category || 'uncategorized',
      tags: metadata.tags || [],
      risk: metadata.risk || 'unknown',
      platforms: metadata.platforms || [],
      triggers: metadata.triggers || []
    });
    
    console.log(`✅ ${skillName} v${metadata.version}`);
  } catch (e) {
    console.error(`❌ Error parsing ${skillName}/SKILL.md: ${e.message}`);
  }
});

// Sort by name
skills.sort((a, b) => a.name.localeCompare(b.name));

// Create index
const index = {
  version: '1.0.0',
  generated: new Date().toISOString(),
  skills
};

// Write index
const indexPath = path.join(__dirname, '../skills_index.json');
fs.writeFileSync(indexPath, JSON.stringify(index, null, 2));

console.log(`\n✅ Generated skills_index.json with ${skills.length} skills`);
console.log(`📄 File: ${indexPath}`);

File: scripts/generate-skills-index.py
#!/usr/bin/env python3
"""Generate skills_index.json from all skills, reading metadata from README.md and SKILL.md."""

import os
import json
import yaml
import re
from datetime import datetime
from pathlib import Path

print("🔍 Scanning skills...")

skills_dir = Path(__file__).parent.parent / 'skills'
skills = []

def extract_from_readme(readme_path):
    """Extract metadata from the README.md table."""
    metadata = {}
    if not readme_path.exists():
        return metadata
    
    with open(readme_path, 'r') as f:
        content = f.read()
    
    for line in content.split('\n'):
        if '|' in line:
            parts = [p.strip() for p in line.split('|') if p.strip()]
            if len(parts) >= 2:
                field = parts[0].lower()
                value = parts[1]
                if field == 'version': metadata['version'] = value
                elif field == 'author': metadata['author'] = value
                elif field == 'category': metadata['category'] = value
                elif field == 'risk': metadata['risk'] = value
                elif field == 'platforms': metadata['platforms'] = [p.strip() for p in value.split(',')]
                elif field == 'tags': metadata['tags'] = [t.strip() for t in value.split(',')]
            
    return metadata

# Read each skill's SKILL.md and README.md
for skill_name in sorted(os.listdir(skills_dir)):
    skill_path = skills_dir / skill_name
    
    if not skill_path.is_dir():
        continue
    
    skill_md = skill_path / 'SKILL.md'
    readme_md = skill_path / 'README.md'
    
    if not skill_md.exists():
        print(f"⚠️  {skill_name}/SKILL.md not found")
        continue
    
    # 1. Start with metadata from README
    meta = extract_from_readme(readme_md)
    
    # 2. Layer on metadata from SKILL.md frontmatter
    with open(skill_md, 'r') as f:
        content = f.read()
    
    if content.startswith('---'):
        lines = content.split('\n')
        yaml_end = None
        for i in range(1, len(lines)):
            if lines[i] == '---':
                yaml_end = i
                break
        
        if yaml_end is not None:
            yaml_content = '\n'.join(lines[1:yaml_end])
            try:
                fm = yaml.safe_load(yaml_content)
                if fm:
                    # YAML frontmatter takes precedence if fields exist
                    for key in ['name', 'description', 'version', 'category', 'risk', 'platforms', 'tags', 'triggers']:
                        if key in fm:
                            meta[key] = fm[key]
            except Exception as e:
                print(f"⚠️ Error parsing YAML in {skill_name}: {e}")

    # Ensure defaults
    skill_data = {
        'name': meta.get('name', skill_name),
        'version': meta.get('version', '1.0.0'),
        'description': meta.get('description', ''),
        'category': meta.get('category', 'general'),
        'tags': meta.get('tags', []),
        'risk': meta.get('risk', 'safe'),
        'platforms': meta.get('platforms', []),
        'triggers': meta.get('triggers', [])
    }
    
    skills.append(skill_data)
    print(f"✅ {skill_name} v{skill_data['version']}")

# Sort by name
skills.sort(key=lambda x: x['name'])

# Create index
index = {
    'version': '1.0.0',
    'generated': datetime.now().isoformat() + 'Z',
    'skills': skills
}

# Write index
index_path = Path(__file__).parent.parent / 'skills_index.json'
with open(index_path, 'w') as f:
    json.dump(index, f, indent=2)

print(f"\n✅ Generated skills_index.json with {len(skills)} skills")
print(f"📄 File: {index_path}")
